### PR TITLE
object/search: Support numeric queries

### PR DESCRIPTION
**Pull request by cthulhu-rider**
* based on and blocked by https://github.com/nspcc-dev/neofs-sdk-go/pull/550

WIP:
 - [x] prelim unit tests
 - [x] support in metabase
 - [x] support in CLI

all ready, waiting for SDK merge

### CODE REVIEW

File: pkg/services/object/search/search.go [hunk 1]
```diff
@@ -48,3 +63,19 @@ func (exec *execCtx) analyzeStatus(execCnr bool) {
 		exec.analyzeStatus(false)
 	}
 }
+
+func verifyQuery(prm Prm) error {
+	for i := range prm.filters {
+		switch prm.filters[i].Operation() {
+		case object.MatchNumGT, object.MatchNumGE, object.MatchNumLT, object.MatchNumLE:
+			// TODO: big math takes less code but inefficient
```
Comment (cthulhu-rider) [inline pkg/services/object/search/search.go:71]: optimization, suggest to defer in separate issue

File: pkg/local_object_storage/metabase/select.go [hunk 2]
```diff
@@ -265,20 +273,75 @@ func (db *DB) selectFromFKBT(
...
+		// TODO: big math takes less code but inefficient
+		filterNum, ok := new(big.Int).SetString(f.Value(), 10)
+		if !ok {
+			db.log.Debug("unexpected non-decimal numeric filter", zap.String("value", f.Value()))
+			return
+		}
+
+		var objNum big.Int
+
+		err := fkbtRoot.ForEach(func(objVal, _ []byte) error {
```
Comment (roman-khimov) [inline pkg/local_object_storage/metabase/select.go:305]: Slow. IIUC it can't be other way with the current DB scheme though.

**Review by roman-khimov — APPROVED** [review]

File: pkg/local_object_storage/metabase/select.go [hunk 3]
```diff
@@ -265,20 +273,75 @@ func (db *DB) selectFromFKBT(
...
-) { //
-	matchFunc, ok := db.matchers[f.Operation()]
-	if !ok {
-		db.log.Debug("missing matcher", zap.Uint32("operation", uint32(f.Operation())))
+) {
+	var nonNumMatcher matcher
+	op := f.Operation()
 
-		return
+	isNumOp := op == object.MatchNumGT || op == object.MatchNumGE || op == object.MatchNumLT || op == object.MatchNumLE
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/select.go:280]: sdk candidate functionality?

File: pkg/local_object_storage/metabase/select.go [hunk 4]
```diff
@@ -265,20 +273,75 @@ func (db *DB) selectFromFKBT(
...
+	if !isNumOp {
+		var ok bool
+		nonNumMatcher, ok = db.matchers[f.Operation()]
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/select.go:283]: `f.Operation()` result is already known

File: pkg/local_object_storage/metabase/select.go [hunk 5]
```diff
@@ -265,20 +273,75 @@ func (db *DB) selectFromFKBT(
...
+		if !ok {
+			db.log.Debug("missing matcher", zap.Uint32("operation", uint32(f.Operation())))
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/select.go:285]: here too

File: pkg/local_object_storage/metabase/select.go [hunk 6]
```diff
@@ -473,6 +536,66 @@ func (db *DB) matchSlowFilters(tx *bbolt.Tx, addr oid.Address, f object.SearchFi
...
+			}
+
+			if filterVal[0] == '-' {
+				if op == object.MatchNumLT || op == object.MatchNumLE {
+					return false
+				}
+				continue
+			}
+
+			if len(filterVal) > 20 { // max uint64 strlen
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/select.go:559]: what if it is 20 len but bigger than uint64? do we return an error for this? even if that is a correct comparison?

File: pkg/local_object_storage/metabase/select.go (hunk 2 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go:305]: i didn't care much bout performance in initial implementation. Left #2757

File: pkg/local_object_storage/metabase/select.go (hunk 3 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go:280]: i wouldn't vote for it

File: pkg/local_object_storage/metabase/select.go (hunk 4 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go]: fixed

File: pkg/local_object_storage/metabase/select.go (hunk 5 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go]: fixed

File: pkg/local_object_storage/metabase/select.go (hunk 6 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go:559]: that was a mistake, fixed

File: pkg/local_object_storage/metabase/select.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/local_object_storage/metabase/select.go:280]: Not a bad idea.

File: pkg/local_object_storage/metabase/select.go [hunk 7]
```diff
@@ -473,6 +536,72 @@ func (db *DB) matchSlowFilters(tx *bbolt.Tx, addr oid.Address, f object.SearchFi
 	}
 
 	for i := range f {
+		switch op := f[i].Operation(); op {
+		case object.MatchNumGT, object.MatchNumGE, object.MatchNumLT, object.MatchNumLE:
```
Comment (roman-khimov) [inline pkg/local_object_storage/metabase/select.go:540]: ...can help avoid `switch` here (`if op.IsNumeric()`)

File: pkg/local_object_storage/metabase/select.go [hunk 8]
```diff
@@ -514,7 +643,39 @@ func groupFilters(filters object.SearchFilters) (filterGroup, error) {
 	}
 
 	for i := range filters {
-		switch filters[i].Header() {
+		hdr := filters[i].Header()
+		//nolint:exhaustive
+		switch filters[i].Operation() {
+		case object.MatchNumGT, object.MatchNumGE, object.MatchNumLT, object.MatchNumLE:
```
Comment (roman-khimov) [inline pkg/local_object_storage/metabase/select.go:649]: ...and here...

File: pkg/services/object/search/search.go [hunk 9]
```diff
@@ -48,3 +63,20 @@ func (exec *execCtx) analyzeStatus(execCnr bool) {
 		exec.analyzeStatus(false)
 	}
 }
+
+func verifyQuery(prm Prm) error {
+	for i := range prm.filters {
+		//nolint:exhaustive
+		switch prm.filters[i].Operation() {
+		case object.MatchNumGT, object.MatchNumGE, object.MatchNumLT, object.MatchNumLE:
```
Comment (roman-khimov) [inline pkg/services/object/search/search.go:71]: and here

File: pkg/local_object_storage/metabase/select.go (hunk 7 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go:540]: or local func `if isNumericOp(op)`

Comment (roman-khimov) [reply pkg/local_object_storage/metabase/select.go:540]: Can be. Let's do this and then it can be moved to SDK if needed anywhere else.

Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/select.go]: done

**Review by carpawell — APPROVED** [review]

**Pull request merged**